runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -345,66 +346,96 @@ MM_IndexableObjectAllocationModel::getSparseAddressAndDecommitLeaves(MM_Environm
-	MM_AllocationContextBalanced *commonContext = (MM_AllocationContextBalanced *)env->getCommonAllocationContext();
 
 	Trc_MM_getSparseAddressAndDecommitLeaves_Entry(envBase->getLanguageVMThread(), spine, (void *)bytesRemaining, bytesRemaining / regionSize, (void *)regionSize);
 	uintptr_t arrayReservedRegionCount = 0;
 	uintptr_t fraction = 0;
 	while (0 < bytesRemaining) {
-		/* Allocate the next arraylet leaf - leaves are allocated solely for the purpose of
+		/* Allocate the next reserved region - reserved regions are allocated solely for the purpose of
 		   decommitting the memory later on in this function. */
 		void *reservedAddressLow = NULL;
-		bool shouldAllocateReservedRegion = true;
+		/* We also remember from which AC the reserved region came from, so that we can return it back to the exact AC when the array is recycled. */
+		MM_AllocationContextBalanced *acForReservedRegion = NULL;
+
 		if (regionSize > bytesRemaining) {
 			fraction = bytesRemaining;
-			/* For code simplicity and lower fragmentation, we always use Common Context. */
-			shouldAllocateReservedRegion = commonContext->allocateFromSharedArrayReservedRegion(envBase, fraction);
-		}
-		if (shouldAllocateReservedRegion) {
-			reservedAddressLow = envBase->_objectAllocationInterface->allocateArrayletLeaf(
+			reservedAddressLow = env->getAllocationContext()->allocateFromSharedReservedRegion(envBase, &_allocateDescription, fraction, true);
+		} else {
+		 	reservedAddressLow = env->_objectAllocationInterface->allocateArrayletLeaf(
 					envBase, &_allocateDescription, _allocateDescription.getMemorySpace(), true);
+		}
 
-			/* If reservedRegion allocation failed set the result to NULL and return. */
-			if (NULL == reservedAddressLow) {
-				Trc_MM_allocateAndConnectNonContiguousArraylet_leafFailure(envBase->getLanguageVMThread());
-				_allocateDescription.setSpine(NULL);
-				spine = NULL;
-				break;
-			}
+		/* Allocation might be fulfilled by another AC */
+		acForReservedRegion = (MM_AllocationContextBalanced *)_allocateDescription.getAllocationContext();
+
+		if (NULL != acForReservedRegion) {
+			reservedRegionAllocationContexts[arrayReservedRegionCount] = acForReservedRegion;
+		} else {
+			Assert_MM_true(NULL == reservedAddressLow);
+			/* reservedRegion allocation failed set the result to NULL and return. */
+			Trc_MM_getSparseAddressAndDecommitLeaves_reserveFailed(envBase->getLanguageVMThread(), arrayReservedRegionCount);
+			fraction = 0;
+			_allocateDescription.setSpine(NULL);
+			spine = NULL;
+			break;
+		}
+
+		if (NULL != reservedAddressLow) {
 			/* Disable region for reads and writes, since accessing virtualLargeObjectHeapAddress through DataAddrForContiguous */
 			void *reservedAddressHigh = (void *)((uintptr_t)reservedAddressLow + regionSize);
 			bool ret = extensions->heap->decommitMemory(reservedAddressLow, regionSize, reservedAddressLow, reservedAddressHigh);
 			if (!ret) {
```

Comment:
these 2 lines could be merged 
intro of ret does not seem to make any line shorter and not seems to be used later